datafusion/core/src/physical_plan/joins/symmetric_hash_join.rs

```diff
@@ -399,10 +393,14 @@ impl ExecutionPlan for SymmetricHashJoinExec {
         self.schema.clone()
     }
 
-    fn required_input_ordering(&self) -> Vec<Option<&[PhysicalSortExpr]>> {
+    fn required_input_ordering(&self) -> Vec<Option<Vec<PhysicalSortRequirement>>> {
```

Comment:
I actually spent some time trying to test bringing this change to IOx and I found this particular part quite challenging:
1. Figuring out `Vec<Option<Vec<..>>>` and how to 
2. `make_sort_requirements_from_exprs` appears not to be public so I couldn't use it.

If we are going to change the signature of these traits, I think we should think about how make it easier to use.

For example,  I wonder if we could make a structure like `Distribution` to https://docs.rs/datafusion/20.0.0/datafusion/physical_plan/enum.Distribution.html to describe an `Order`? 


This would both 
1. make it more convenient to construct these lists 
2. provide a place to document what they are and all the subtlety 

Perhaps something like 

```rust
impl ExecutionPlan {
    // the order required for the plans inputs
    fn required_input_ordering(&self) -> Vec<Ordering> ;

    // output order produced by this plan
    fn output_ordering(&self) -> Ordering ;
}
```

And then encapsulate all the gory details about orderings and comparing them in `Ordering`:

```rust 
struct Ordering {
  exprs: Vec<dyn PhysicalExpr>
  // parallel list if there are any particular requirements on the sort orders
  required_options: Vec<SortOptions>
}

impl RequiredInputOrdering {
  // construct a required input 
  pub fn from_sort_exprs(exprs: Vec<dyn PhysicalSortExpr>) -> Self {
  .. 
  }
 /// does this represent any requirement on the input ordering
 fn input_order_required(&self) ->  bool {}

 // Is there any NULLS FIRST / NULLS LAST 
...
}
```

Reply:
I think you mean something like:
```rust
type Ordering = Vec<PhysicalSortExpr>;
type OrderingRequirement = Vec<PhysicalSortRequirement>;
```
and which would allow us increase readability by changing
```rust
impl ExecutionPlan {
    fn required_input_ordering(&self) -> Vec<Option<Vec<PhysicalSortRequirement>>>;
    fn output_ordering(&self) -> Option<&[PhysicalSortExpr]>;
}
```
to
```rust
impl ExecutionPlan {
    fn required_input_ordering(&self) -> Vec<Option<OrderingRequirement>>;
    fn output_ordering(&self) -> Option<Ordering>;
}
```
Am I understanding your idea right?

Reply:
We can also create new outright types instead of using aliases, as in:
```rust
struct Ordering(Vec<PhysicalSortExpr>);
struct OrderingRequirement(Vec<PhysicalSortRequirement>);
```
We would need some boilerplate to make them ergonomic when using elsewhere; i.e.
```rust
impl core::ops::Deref for Ordering {
    type Target = Vec<PhysicalSortExpr>;

    fn deref (self: &'_ Self) -> &'_ Self::Target {
        &self.0
    }
}

impl core::ops::DerefMut for Ordering {
    fn deref_mut (self: &'_ mut Self) -> &'_ mut Self::Target {
        &mut self.0
    }
}

impl core::ops::Deref for OrderingRequirement {
    type Target = Vec<PhysicalSortRequirement>;

    fn deref (self: &'_ Self) -> &'_ Self::Target {
        &self.0
    }
}

impl core::ops::DerefMut for OrderingRequirement {
    fn deref_mut (self: &'_ mut Self) -> &'_ mut Self::Target {
        &mut self.0
    }
}
```
which will allow us to avoid a public `make_sort_requirements_from_exprs` and instead offer:
```rust
impl From<Ordering> for OrderingRequirement {
    fn from(value: Ordering) -> Self {
        OrderingRequirement(make_sort_requirements_from_exprs(&value))
    }    
}
```

If I get the spirit of your suggestion right, we can experiment with these and incorporate such a change in our next commit addressing the feedback. We can explore aliasing vs. tuple-struct methods and share our findings along with the commit.

Reply:
OK, having played with these ideas for 2-3 hours, here are my observations:

Changing the return types of `output_ordering` and `required_input_ordering` with non-plug-in-compatible types are really challenging. I made two refactor attempts: One with a new `struct` and one with a tuple-struct like my previous example. I couldn't finish either of them in 2-3 hours -- it turns out doing this results in massive changes and things are quite intertwined.

However, I have some good news too: We can achieve an almost-as-good readability level with type aliasing:
```rust
pub type ExprOrdering = Vec<PhysicalSortExpr>;
pub type ExprOrderingRef<'a> = &'a [PhysicalSortExpr];
pub type OrderingRequirement = Vec<PhysicalSortRequirement>;
```
With these "types", I was able complete the refactor and significantly simplify how the code reads; i.e.
```rust
impl ExecutionPlan {
    fn required_input_ordering(&self) -> Vec<Option<OrderingRequirement>>;
    fn output_ordering(&self) -> Option<ExprOrderingRef>;
}
```
with `ExprOrdering` replacing `Vec<PhysicalSortExpr>` where the actual vector is required.

This refactor is not large in terms of LOCs (~250 lines change, almost entirely consisting of simple substitutions), but it touches almost 50 files! So we may want to leave it to a follow-on too -- I am fine with either.